bsp/Src/UART.c

```diff
@@ -0,0 +1,429 @@
+    for (uint8_t i = 0; i < length; i++) {
+        if (xQueueSend(*tx_queue, &data[i], timeout) != pdTRUE) {
+            status = UART_ERR;
+            goto exit;
+        }
+    }
+
+    // Start transmission if needed
+    if (HAL_UART_GetState(handle) != HAL_UART_STATE_BUSY_TX) {
+        uint8_t byte;
+        if (xQueueReceive(*tx_queue, &byte, 0) == pdTRUE) {
+            if (HAL_UART_Transmit_IT(handle, &byte, 1) != HAL_OK) {
+                status = UART_ERR;
+                goto exit;
+            }
+        }
+    }
+
+exit:
+    xSemaphoreGive(*send_semaphore);
+    return status;
+}
+
+/**
+ * @brief Reads data from the user-provided RX queue
+ * @param huart pointer to the UART handle
+ * @param data pointer to the data buffer that will be read into
+ * @param length of the buffer that will be read
+ * @param blocking if true, function will block until data is received
+ * @return uart_status_t
+ */
+uart_status_t uart_recv(UART_HandleTypeDef* handle, uint8_t* data, uint8_t length, bool blocking) {
+    if (!initialized || !data || length == 0) {
+        return UART_ERR;
+    }
+
+    QueueHandle_t* rx_queue = NULL;
+    if(handle->Instance == UART4) {
+        rx_queue = uart4_rx_queue;
+    }
+    #ifdef UART5
+    else if(handle->Instance == UART5) {
+        rx_queue = uart5_rx_queue;
+    }
+    #endif /* UART5 */
+    else {
+        return UART_ERR;
+    }
+
+    if (!rx_queue) {
+        return UART_ERR;
+    }
+
+    TickType_t timeout = blocking ? portMAX_DELAY : 0;
+    uart_status_t status = UART_RECV;
+
+    for (uint8_t i = 0; i < length; i++) {
+        if (xQueueReceive(*rx_queue, &data[i], timeout) != pdTRUE) { // extract data from the rx queue and store it in data
+            status = blocking ? UART_ERR : UART_EMPTY;
+            return status;
+        }
+    }
+
+    return status;
+}
+
+// Transmit Callback occurs after a transmission if complete (depending on how huart is configure)
+void HAL_UART_TxCpltCallback(UART_HandleTypeDef *huart) {
+    BaseType_t higherPriorityTaskWoken = pdFALSE;
+    uint8_t nextByte;
+    QueueHandle_t* tx_queue = NULL;
+
+    if(huart->Instance == UART4) {
+        tx_queue = &uart4_tx_queue;
+    }
+    #ifdef UART5
+    else if(huart->Instance == UART5) {
+        tx_queue = &uart5_tx_queue;
+    }
+    #endif /* UART5 */
+    else {
+        return;
+    }
+
+    if (xQueueReceiveFromISR(*tx_queue, &nextByte, &higherPriorityTaskWoken) == pdTRUE) {
+        HAL_UART_Transmit_IT(huart, &nextByte, 1);
+    }
+    
+    portYIELD_FROM_ISR(higherPriorityTaskWoken);
+}
+
+// Receive Callback occurs after a receive is complete
+void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart) {
+    QueueHandle_t* rx_queue = NULL;
+    if(huart->Instance == UART4) {
+        rx_queue = uart4_rx_queue;
+    }
+    #ifdef UART5
+    else if(huart->Instance == UART5) {
+        rx_queue = uart5_rx_queue;
+    }
+    #endif /* UART5 */
+    else {
+        return;
+    }
```

Comment:
Not needed since the passed in arguement is from the driver, not the application